src/Symfony/Component/HttpClient/Response/AsyncResponse.php

```diff
@@ -0,0 +1,355 @@
+            return;
+        }
+
+        $context = new AsyncContext($this->passthru, $client, $this->response, $this->info, $this->content, $this->offset);
+        if (null === $stream = ($this->passthru)(new LastChunk(), $context)) {
+            return;
+        }
+
+        if (!$stream instanceof \Iterator) {
+            throw new \LogicException(sprintf('A chunk passthru must return an "Iterator", "%s" returned.', get_debug_type($stream)));
+        }
+
+        try {
+            foreach ($stream as $chunk) {
+                if ($chunk->isLast()) {
+                    break;
+                }
+            }
+
+            $stream->next();
+
+            if ($stream->valid()) {
+                throw new \LogicException('A chunk passthru cannot yield after the last chunk.');
+            }
+
+            $stream = $this->passthru = null;
+        } catch (ExceptionInterface $e) {
+            // ignore any errors when canceling
+        }
+    }
+
+    /**
+     * @internal
+     */
+    public static function stream(iterable $responses, float $timeout = null, string $class = null): \Generator
+    {
+        while ($responses) {
+            $wrappedResponses = [];
+            $asyncMap = new \SplObjectStorage();
+            $client = null;
+
+            foreach ($responses as $r) {
+                if (!$r instanceof self) {
+                    throw new \TypeError(sprintf('"%s::stream()" expects parameter 1 to be an iterable of AsyncResponse objects, "%s" given.', $class ?? static::class, get_debug_type($r)));
+                }
+
+                if (null !== $e = $r->info['error'] ?? null) {
+                    yield $r => $chunk = new ErrorChunk($r->offset, new TransportException($e));
+                    $chunk->didThrow() ?: $chunk->getContent();
+                    continue;
+                }
+
+                if (null === $client) {
+                    $client = $r->client;
+                } elseif ($r->client !== $client) {
+                    throw new TransportException('Cannot stream AsyncResponse objects with many clients.');
+                }
+
+                $asyncMap[$r->response] = $r;
+                $wrappedResponses[] = $r->response;
+            }
+
+            foreach ($client->stream($wrappedResponses, $timeout) as $response => $chunk) {
+                $r = $asyncMap[$response];
+
+                if (!$r->passthru) {
+                    if (null !== $chunk->getError() || $chunk->isLast()) {
+                        unset($asyncMap[$response]);
+                    }
+
+                    yield $r => $chunk;
+                    continue;
+                }
+
+                $context = new AsyncContext($r->passthru, $r->client, $r->response, $r->info, $r->content, $r->offset);
+                if (null === $stream = ($r->passthru)($chunk, $context)) {
+                    if ($r->response === $response && (null !== $chunk->getError() || $chunk->isLast())) {
```

Comment:
We are doing `$asyncMap[$r->response] = $r;` so doing `$r->response === $response` is not necessary, is it?

Reply:
Yes: AsyncContext can change the response property (private communication protocol based on references, same as for passthru and info)